### DISCUSSION TITLE
Tasmota UFS - universal file system documentation

### DISCUSSION THREAD

**Discussion by rapi3** [arendst/Tasmota#11617] [General]
Hello,
As promised with your help I prepared a "small" documentation for UFS - universal file system to be included in Tasmota doc.
Please review it and let me know _where can be added_ or just add it there.

`**Tasmota UFS - universal file system**

Since a ESP8266-12F has a 4MB flash you can build a variant with a filesystem and store there files, data, commands.
The file partition does NOT get erased by reset commands or factory default. Only a flash erase it will wipe it.
There are some special files that you can upload and use:

**autoexec.bat**
It is a special file that will store commands that will be executed at every boot.
Almost any commands can be used.
Avoid any command that will make the device to reboot like changing Wifi setting, MQTT settings, Templates & GPIO...
otherwise it will reboot continuously (in fact 3 or 4 times then autoexec will be disabled after boot loop detection)
Every command in a extra line in autoexec.bat will be executed sequentially.

**display.ini**
It is a special file that will store data to be displayed at every boot if you have a display attached.
example:
; clr screen
[z]
; draw full screen picture - corona.rgb file must exist in UFS storage
[x0y0P/corona.rgb:]
; define index color
[dc19:31000]
; draw transparent text with new index color
[x60y30f2Ci19D2]Tasmota

**UFS commands** can be found here:
[https://tasmota.github.io/docs/Commands/#ufs](url)

**Berry scripting language** can benefit of UFS, more info here:
[https://tasmota.github.io/docs/Berry-Scripting/#loading-code-from-filesystem](url)

**How to activate UFS ?**
You need to compile your firmware to activate this feature:
Copy platformio_override_sample.ini as platformio_override.ini
Activate/uncomment the line: board_build.ldscript = eagle.flash.4m2m.ld

Add or uncomment #define lines in your user_config_override.h:
// UFS filesystem
#define USE_UFILESYS
  #define GUI_TRASH_FILE

Compile, flash and you will find a new entry in Tasmota web GUI menu:
Configuration - Manage File system.`
![fs-1](https://user-images.githubusercontent.com/18531150/113870887-9b469880-97b2-11eb-91f2-ea129fce6427.jpg)
![fs-2](https://user-images.githubusercontent.com/18531150/113870899-9e418900-97b2-11eb-8814-0435060cfb6c.jpg)

**Comment by barbudor** [comment]
Nice start @rapi3 
It may be worth a dedicated page in the Docs
Proper formatting will be required but there are plenty of sample in the current doc on how to do this. And the doc repository has some guidelines on how to do this : https://github.com/tasmota/docs/blob/development/README.md

**Reply by sfromis** [reply]
If would be relevant to also link to the UFS commands starting here:
https://tasmota.github.io/docs/Commands/#ufs

Maybe also more references to where Tasmota features already now are based on having the file system, including the Berry scripting language, a new feature for ESP32
https://tasmota.github.io/docs/Berry-Scripting/#loading-code-from-filesystem

**Reply by rapi3** [reply]
updated from your suggestion

**Comment by rapi3** [comment]
I think I need to add some clarification:
When exactly autoexec.bat and display.ini it is executed:
After boot - after Tasmota is fully intialized - after Wi-Fi and MQTT (if enabled) are connected ?

**Reply by sfromis** [reply]
`autoexec.bat` is a simple sequence of commands (backlog-like) executed after MQTT initialization, connected or not, basically after the /INFO3 payload, similar to the rules trigger `System#Boot`. They should not contain anything causing restart.

**Reply by sfromis** [reply]
`display.ini` has a list of DiplayText commands, one per line, executed at display driver initialization time, which AFAIK is much earlier in the boot process, before Wifi and MQTT, similar to the rules trigger `System#Init`.

**Reply by rapi3** [reply]
ok thank you I updated
https://github.com/tasmota/docs/pull/703

**Comment by nikito7** [comment]
Is file upload working?

I have issues with files over 4kb

**Reply by sfromis** [reply]
"Issues" ... why keep details a secret? Someone might have something to add.

Anyway, I just uploaded a file of approx 100KB on an ESP32 device without isses; could download the image file again, and display it.

**Reply by nikito7** [reply]
In esp8266

![Screenshot_20220922-235118_Chrome](https://user-images.githubusercontent.com/45373783/191865080-ac2c2073-fec4-4e91-aead-233b7267df89.jpg)

**Comment by canufarm** [comment]
Hi,
how can I upload a html file and then use this anywhere in the network to be loaded? When I click on the filename link the file gets downloaded with download content type. Any chance to use the file system as "webspace"? Thanks.

**Reply by nikito7** [reply]
/ufs/index.html

**Reply by canufarm** [reply]
Thank you, I did already try this, but I only get:

```
File Not Found

URI: /ufs/test.html
Method: GET
Arguments: 0
```

**Reply by nikito7** [reply]
It works here. Esp8266. Custom build.

![Screenshot_20221203-182830_Chrome](https://user-images.githubusercontent.com/45373783/205456298-e7f89eec-1f34-4671-be5b-19a02db47057.jpg)

![Screenshot_20221203-182900_Chrome](https://user-images.githubusercontent.com/45373783/205456313-b0cf4fc5-5ea7-4dd1-9aaf-a1dfec57c662.jpg)

**Reply by canufarm** [reply]
Hm, I use official build with ESP32 Cam.

**Reply by sfromis** [reply]
For me, `http://tasmotadevice/ufs/testfile.txt` neither works on ESP32 nor ESP8266 builds with file system. Did you add anything nonstandard to that custom build?

**Reply by sfromis** [reply]
Anyway, here's an example of a very rudimentary web server created using the Berry language, serving URLs like `http://tasmotadevice/fs?file=testfile.txt`
```
class BerryFS
  def init()
    tasmota.add_driver(self)
  end
  def webrequest()
    import webserver
    try
      var fh = open(webserver.arg('file'),'r')
      webserver.content_response(fh.read())
      fh.close()
    except .. as e, m
      webserver.content_response('error:'..e..' '..m)
    end
  end
  def web_add_handler()
    import webserver
    webserver.on("/fs", / -> self.webrequest(), webserver.HTTP_GET)
  end
end
var berryfs = BerryFS()
```

**Reply by nikito7** [reply]
> For me, `http://tasmotadevice/ufs/testfile.txt` neither works on ESP32 nor ESP8266 builds with file system. Did you add anything nonstandard to that custom build?

With tasmota32-webcam dont work.

I have scripts and sml stuff

**Reply by nikito7** [reply]
Working with:

https://github.com/nikito7/easyhan/blob/dev/edpbox/esp32/32-12.2.0-5.2.zip

Or:

One of them 😆

```
// ### ### ###
// user_config_override.h
// sml configs

#undef USE_RULES
#define USE_SCRIPT
#define USE_SML_M
#define USE_SML_SCRIPT_CMD
#define SML_MAX_VARS 50
#define SML_REPLACE_VARS

// tcp

#define USE_TCP_BRIDGE
#define USE_MODBUS_BRIDGE_TCP
#define USE_MODBUS_BRIDGE

// misc

#define USE_SCRIPT_WEB_DISPLAY
#define MAX_METERS 1
#define MAXVARS 55
#define MAXSVARS 20
#define MAXFILT 10
#define LARGE_ARRAYS
#define SCRIPT_LARGE_VNBUFF
#define USE_GOOGLE_CHARTS
#define USE_SCRIPT_JSON_EXPORT
// 
// #define USE_SCRIPT_SUB_COMMAND // UfsRun issues?!

// eof
```

**Reply by sfromis** [reply]
Not available without scripting, thus not a general feature.
```
#if defined(USE_UFILESYS) && defined(USE_SCRIPT_WEB_DISPLAY)
      Webserver->on(UriGlob("/ufs/*"), HTTP_GET, ScriptServeFile);
#endif
```

**Comment by evidalr** [comment]
By mistake, I have tried to upload a too-large file and after a while, the upload process has failed saying "file too large" or something like that.   After that incident, the UFS partition appears to be corrupt, because the console command UfsFree says: 'RESULT = {"UfsFree":1968}' (also as displayed in the the Manage File system wirndow), but if I try to upload anything I get: "Upload Failed / Not enough space".

I havte tried the UfsDelete command in the console, but I get:  'RESULT = {"Command":"Error","Input":"UFSDELETE"}'.

Any idea as to how to recover the UFS "lost" space? -- Thanks in advance!

**Reply by sfromis** [reply]
You are not clear about what exactly you do to upload. IIRC, that "Not enough space" message may mean nothing more than an error occurred, not limited to space issues.

And the `UfsDelete` command output you posted is indeed correct (as in the command being in error). You need a file name to delete a file.

**Reply by evidalr** [reply]
Thanks for the feedback!
I can post more details this night.  But in any case, the net effect is that the Tasmota UFS partition is apparently almost empty ({"UfsFree":1968}), but any attempt to upload a file (even a small one of a few KB) results in: "Upload Failed / Not enough space".  

Looks like the filesystem was corrupted as a result of trying to upload a very big file.  So the question remains:   Any idea as to how to recover the UFS "lost" space?  or, what is the procedure to completely erase (and re-format) the UFS partition? Would a full recovery make it?

Thanks again in advance for any help.

**Reply by sfromis** [reply]
Again, the message "Not enough space" can occur with other errors, which also depends on what method you are using to upload. Hence, assuming that the file system is corrupted does not follow from the few details you posted. Anyway, I'd retry after restarting. Or check what happens if using the web UI to create a file (not upload).

I'm not seeing trouble after trying to upload a file too large for available free space.

One approach to getting a fresh file system might be to use the Partition Wizard to resize it, loosing content.

**Reply by evidalr** [reply]
Thanks again for your patience.  After testing what you propose and other tests, now I confirm your guess that the message "Not enough space" does **not** actually mean that there is not space left!  

The problem happens only when trying to upload (using the web interface) a specific file; namely: "Conf_tasmota_40CCA9_3241_14.5.0.dmp" (which has only 4096 bytes).  It seems related with the file **name**, because if I copy this file to "test.dmp", it is uploaded flawlessly. I have also been able to upload other large files, even one also with a complicated name.

Interestingly, in another Tasmota device, I have uploaded without problems the similar tile "ConfTasm_F3B342_4930_14.5.0.dmp".

It is really strange, but inot a big problem anymore.  I was fearing filesystem corruption, which would actually had been a problem...

Thanks again for your help.

**Reply by evidalr** [reply]
It could be a filename length issue.  I have uploaded it without problem with the new name: "confTasm0CCA9-3241-14.5.0.dmp".

All the best,

**Reply by sfromis** [reply]
Uploading a configuration backup file to the Tasmota file system is a "pretty odd" thing to do, as those are meant to be used with configuration backup/restore. I'm not making a guess about this explaining what you saw (or not), just giving a bit of context.

**Reply by evidalr** [reply]
sfromis wrote:
> Uploading a configuration backup file to the Tasmota file system is a "pretty odd"

Yes, I know ;)  I was just testing different usage possibilities for the UFS.  Sotoring there an additional backup copy of the Tasmota config backup file, along with a plaintext file with all the rules currently implemented in the device  was an idea....

**Reply by evidalr** [reply]
Just to close this issue:

I have systematically tried to upload empty files with random names from 1 up to 32 charaters long.  All the uploads succeeded up to length 31.  Those with length 32 failed systematically with "Upload Failed / Not enough space".  

The message "Not enough space" is a bug (or at least is misleading).
I do not know how to fix it myself (I guess in the Tasmota source files).

The length limit should be documented in https://tasmota.github.io/docs/UFS.
I have just posted a pull request to add "*Beware, only filenames up to 31 characters long are allowed.**" to the description of the Upload button.

**Reply by sfromis** [reply]
Arguably, "not enough space" is a correct message, when you are trying to use names longer than the available space for names :grin:
